### PR TITLE
feat: user-facing part of variable vnode count

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This is a progress towards https://github.com/risingwavelabs/risingwave/issues/15900.

Significant changes in this PR:

- The vnode count for singletons will always be 256 (no matter existing or newly created), instead of 1 used in previous PRs. This is to simplify backwards compatibility.
- Add session variable named `streaming_max_parallelism` to decide the max parallelism used in future streaming jobs. Note that the default value of the session variable can be changed globally with `ALTER SYSTEM`.
- Retrieve correct value of vnode count from expression context in `rw_vnode()`, used internally for 2-phase aggregation.
- Rename `VirtualNode::COUNT` to `VirtualNode::COUNT_FOR_COMPAT`, indicating that it should not be used in most of the code. Also update the Java binding function.
- Other minor fixes of TODOs.

The variable vnode count feature is now generally available after this PR, with just a few individual TODOs that still need to be fixed.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Introduce a new session config of `streaming_max_parallelism`. Update the doc of `streaming_parallelism`. Please refer to the rustdoc for documentation.

### CODE REVIEW

**Comment by BugenZhao** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18515?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#18515** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18515?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#18444** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18444?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18415** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18415?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18466** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/18466?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @BugenZhao and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by fuyufjh — APPROVED** [review]
Generally LGTM!

File: src/meta/src/rpc/ddl_controller.rs [hunk 1]
```diff
@@ -1634,7 +1634,11 @@ impl DdlController {
         let parallelism = self.resolve_stream_parallelism(specified_parallelism, &cluster_info)?;
         let parallelism_limited = parallelism > max_parallelism;
         if parallelism_limited {
-            tracing::warn!("Too many parallelism, use {} instead", max_parallelism);
+            // TODO(var-vnode): may return error here?
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/meta/src/rpc/ddl_controller.rs:1637]: > * if the parallelism is specified by user, returning error is better

Agree. We need to differentiate between these cases.

File: src/meta/src/stream/stream_graph/schedule.rs [hunk 2]
```diff
@@ -152,11 +152,11 @@ impl Distribution {
     }
 
     /// Get the vnode count of the distribution.
-    // TODO(var-vnode): after `ServingVnodeMapping::upsert` is made vnode-count-aware,
-    // we may return 1 for singleton.
+    ///
+    /// For backwards compatibility, [`VirtualNode::COUNT_FOR_COMPAT`] is used for singleton.
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/meta/src/stream/stream_graph/schedule.rs:156]: I don't remember exactly, but I previously encountered some problems and found that using 256 for the singleton simplified things.

However, it's true that using `1` is much clearer and more consistent. Given that this PR has passed the tests, I'll make another attempt in future PRs.

**Pull request merged**